### ISSUE TITLE
RFC: Replace Newtonsoft with System.Text.Json

### ISSUE THREAD

**Issue by adamrodger** [pact-foundation/pact-net#458]
## Introduction

This is a request for comments from the community about a proposal to replace Newtonsoft with System.Text.Json as the
main supported JSON serialiser.

## Background

PactNet 4.x and below use the Newtonsoft JSON serialiser to:

- Serialise matchers to the special internal format required by FFI
- Serialise dynamic request and response bodies for interactions
- Deserialise message interaction contents from the FFI during verification
- Serialise consumer version selectors when loading pacts from a broker during verification

However, PactNet is most likely to be used with ASP.Net Core going forwards, in which the default JSON serialiser is
Microsoft's own System.Text.Json package. This creates friction when working with PactNet because any custom serialisation
options applied to request/response objects need to be reimplemented in Newtonsoft just to work with pact tests.

This reimplementation could be incorrect because the two libraries sometimes behave differently, and therefore the worst
case scenario for PactNet could easily happen - the pact tests will all pass but the service fails in production usage at
runtime.

## Proposal

All references to Newtonsoft JSON serialiser will be removed for PactNet 5.x and replaced with references to the Microsoft
JSON serialiser.

This will create some breaking API changes where we currently support user-supplied serialisation settings, but the proposal
is simply to swap like-for-like from Newtonsoft's `JsonSerializerSettings` to `System.Text.Json.JsonSerializerOptions`.
This will create a minimal impact during migration whilst still allowing us to change serialiser.

Internally, PactNet will use `System.Text.Json` to serialise matchers and dynamic request/response bodies supply by users,
and to deserialise message interaction requests from the FFI during message interaction verification. To the user these
are entirely transparent changes.

The library will continue to ship as `netstandard2.0` instead of changing to a direct .Net version so that existing compatibility
is maintained. `System.Text.Json` supports `netstandard2.0` explicitly.

### Version

There is an open question on exactly which version of STJ to reference. If we pick the current version then we may prevent people from upgrading if they're on older version. If we pick an older version then which one?

### Breaking Serialisation Changes

Any custom serialisation options applied to objects previously which were not being used may now start to be used, and so
potential behavioural differences might happen post-upgrade. For example, given the class:

```csharp
public class MyDto
{
    public int Foo { get; init; }

    [System.Text.Json.Serialization.JsonIgnore]
    public string Bar { get; init; }
}
```

In PactNet 4.x the `Bar` property would still be serialised, whereas following this change it would no longer be serialised.
This behaviour seems advantageous rather than problematic, as it also wouldn't be serialised by the real API under real usage
and thus the pacts were technically using the wrong format.

However, there are situations where the behaviour of serialisation itself may have been altered which could potentially cause
behavioural changes. It's not felt the risk is high or that the consequence is severe in these situations though. The correct
action to perform in that case would ensure that the pacts and API are as close as possible anyway.

**Comment by jvmlet** [comment]
My request as community member  to stay with `Newtonsoft` as default  json serializer.
Majority of projects _I know_  are stick to `nswag` generator that uses  `Newtonsoft` , plus it's more feature rich than MS's one.

**Comment by adamrodger** [comment]
The default JSON serialiser shipped with ASP.Net Core is STJ instead of Newtonsoft, and the default OpenAPI library is Swashbuckle instead of NSwag.

If we choose not to go with those defaults I think we need a good rationale beyond the things we personally prefer.

**Comment by jvmlet** [comment]
I wouldn't expect  asp.net core to ship some  non-MS package as default. 
Swashbuckle is open api _hosting_ middleware, but nswag `generator` produces Newtonsoft annotated dtos.
The fact that some package is developed  by MS  and is shipped by default with asp.core doesn't make it THE  standard...
Newtonsoft is more feature rich comparing to stj, as well as autofac is way more sophisticated as MS's DI container.

**Comment by adamrodger** [comment]
The debate isn't really about which library is better, because PactNet is a testing library.

The debate is which makes PactNet easiest to use for our users. So, given STJ is default in ASP.Net Core and the vast majority of our users will be using ASP.Net Core, the lowest friction thing for the vast majority of users would be to go with that library.

At the moment with PactNet being on Newtonsoft whereas the majority of new code is on STJ, it means users have to do workarounds specifically to support tests. These are things like adding Newtonsoft annotations even though they don't use it in production, or making the test setup very awkward in the pact tests. The aim is to reduce this friction as much as possible.

**Comment by jvmlet** [comment]
I'm not debating which one is better. My point is that  according to _my experience_, developers are using nswag stack to generate and implement web api from open api specs. So they are **already** using Newtonsoft as their json serializer and will have difficulties to support pact expectations as stj serialized json.

**Comment by adamrodger** [comment]
Ok, well your feedback is taken on board, thanks.

**Comment by DavidHvilava** [comment]
Hello) FYI, Newtonsoft.Json version [10.0.3](https://www.nuget.org/packages/Newtonsoft.Json/10.0.3) that's currently used in Pact.net has security vulnerability. It's been fixed in 13.0.1. In a company I work for we use a tool for security compliance that indicated that problem. 
So if switching to System.Text.Json is not feasible in the nearest time maybe you will consider upgrading Newtonsoft.Json. Thanks
![image](https://github.com/pact-foundation/pact-net/assets/77272306/db3bad09-16c3-4449-963c-ceec6c230bdb)

**Comment by adamrodger** [comment]
The version listed is the minimal supported version. PactNet will use a higher version if you have one installed.

**Comment by DavidHvilava** [comment]
Actually, you're right. I can see it now in VS solution explorer. I will check why our tool complains about that. Thanks!

**Comment by adamrodger** [comment]
Merged to be included in `5.0.0-beta.2`

**Comment by andyfurniss4** [comment]
Thanks for this. I'm glad STJ is replacing Newtonsoft. These days I would expect most libraries to use STJ over Newtonsoft.

Do you know when v5 will be fully released?
